Update the implementation details of the external bus for the apps (#3049)
home-assistant/developers.home-assistant: docs/frontend/external-bus.md

Messages are passed to the external app as serialized JSON objects. The function that will be called takes a single parameter: a string. The external app will have to process the message and deal with it accordingly (or ignore it).

On Android, the implementation depends on the WebView features:

**V2 (recommended)**: Uses [`WebViewFeature.WEB_MESSAGE_LISTENER`][web-message-listener] for secure origin validation. The frontend sends messages using:

```ts
postMessage(JSON.stringify({ type: "externalBus", payload: msg }))
```

**V1 (fallback)**: Used when the WebView doesn't support [`WebViewFeature.WEB_MESSAGE_LISTENER`][web-message-listener]. Your app needs to define:

```ts
window.externalApp.externalBus(message: string)
```

:::note
The Android app automatically selects V2 when available (Home Assistant 2026.4.2 or later) and falls back to V1 for compatibility with older servers or WebView versions that don't support the [`WebViewFeature.WEB_MESSAGE_LISTENER`][web-message-listener].
:::

> 

Reviewer: This is an implementation detail for the Home Assistant Companion app on Android and shouldn't be in this documentation

Author: I added it here because we speak about V1/V2 without telling when V2 is used. Maybe I'm overthinking this and it is irrelevant wdty?

Author: Actually it is somehow describe above implicitly since I explain that it requires a feature. I'm removing it.

Messages are passed to the external app as serialized JSON objects. The function that will be called takes a single parameter: a string. The external app will have to process the message and deal with it accordingly (or ignore it).

On Android, the implementation depends on the WebView features:

> **V2 (recommended)**: Uses [`WebViewFeature.WEB_MESSAGE_LISTENER`][web-message-listener] for secure origin validation. The frontend sends messages using the injected V2 object:

Reviewer: Code block at the end seems to be missing

Author: Fixed 👍🏻 thanks for the catch I'm merging the PR now.

Payload: None

Expected response:

```ts
{
  hasSettingsScreen?: boolean;
  hasSidebar?: boolean;
  canWriteTag?: boolean;
  hasExoPlayer?: boolean;
  canCommissionMatter?: boolean;
  canImportThreadCredentials?: boolean;
  canTransferThreadCredentialsToKeychain?: boolean;
  hasAssist?: boolean;
  hasBarCodeScanner?: number;
  canSetupImprov?: boolean;
  appVersion?: string;
  hasEntityAddTo?: boolean;
  hasAssistSettings?: boolean;
}
```

- `hasSettingsScreen`: Set to true if the external app will show a configuration screen when it receives `config_screen/show`
- `hasSidebar`: Set to true if the external app has a sidebar
- `canWriteTag`: Set to true if the external app can write NFC tags (i.e., the device has NFC hardware)
- `hasExoPlayer`: Set to true if the app supports HLS video playback via ExoPlayer
- `canCommissionMatter`: Set to true if the app can commission Matter devices
- `canImportThreadCredentials`: Set to true if the app can import Thread credentials
- `canTransferThreadCredentialsToKeychain`: Set to true if the app can transfer Thread credentials to the keychain (Apple only)

> - `hasAssist`: Set to true if the app supports the Assist voice assistant interface

Reviewer: ```suggestion
- `hasAssist`: Set to true if the app has a native Assist interface, replacing the frontend implementation
```

> - `hasBarCodeScanner`: Set to 1 if the app has barcode scanning capability, 0 otherwise

Reviewer: ```suggestion
- `hasBarCodeScanner`: Set to `1` if the app has bar code scanning capability, `0` otherwise
```

If I remember correctly the idea was the use the number as a version/feature level here so a v2 could be introduced in the future without needing to expand the external bus config.

> - `canSetupImprov`: Set to true if the app can set up Improv-enabled devices

Reviewer: ```suggestion
- `canSetupImprov`: Set to true if the app can set up Improv Wi-Fi devices
```

- `appVersion`: The version string of the native app
- `hasEntityAddTo`: Set to true if the app supports adding entities to platform-specific locations (e.g., homescreen widget)
- `hasAssistSettings`: Set to true if the app has an Assist settings screen

##### `entity/add_to/get_actions`

Get available actions for adding an entity to the device (e.g., homescreen, watch face).

Payload:

```ts
{
  entity_id: string;
}
```

Expected response:

```ts
{
  actions: Array<{
    enabled: boolean;
    name: string;
    details?: string;
    mdi_icon: string;
    app_payload: string;
  }>;
}
```

- `enabled`: Whether the action is currently available
- `name`: Display name of the action
- `details`: Optional additional details about the action
- `mdi_icon`: Material Design Icon identifier for the action (e.g., "mdi:car")
- `app_payload`: Opaque string to be sent back in `entity/add_to` to execute the action

> #### Messages not expecting a response

Reviewer: I understand there was already content here but maybe sort this list a-z to make it easier to scan.

}
```

- `tag`: The tag ID to write
- `name`: The name of the tag as entered by the user, or `null` if no name has been set

##### `assist/show`

> Show the Assist interface.

Reviewer: ```suggestion
Show the native Assist interface.
```

Payload (optional):

```ts
{
  pipeline_id: "preferred" | "last_used" | string;
  start_listening: boolean;
}
```

- `pipeline_id`: The pipeline to use, or "preferred"/"last_used" for automatic selection
- `start_listening`: Whether to start listening immediately

##### `assist/settings`

Open the Assist settings screen.

Payload: None

##### `sidebar/show`

Show the sidebar.

Payload: None

##### `bar_code/scan`

> Start scanning a barcode.

Reviewer: Frontend has a _, so let's use a space. Both spellings seem to be valid.

```suggestion
Start scanning a bar code.
```

Payload:

```ts
{
  title: string;
  description: string;
  alternative_option_label?: string;
}
```

- `title`: Title to show in the scanner UI (required)
- `description`: Description text to show in the scanner UI (required)

> - `alternative_option_label`: Optional label for an alternative action button

Reviewer: ```suggestion
- `alternative_option_label`: Optional label for an alternative action button, if not included the alternative action button should be hidden
```

##### `bar_code/close`

> Close the barcode scanner.

Reviewer: ```suggestion
Close the bar code scanner.
```

Payload: None

##### `bar_code/notify`

> Show a notification message in the barcode scanner.

Reviewer: ```suggestion
Show a notification message in the bar code scanner.
```

Payload:

```ts
{
  message: string;
}
```

##### `matter/commission`

Start the Matter device commissioning flow.

Payload (optional):

```ts
{
  mac_extended_address: string | null;
  extended_pan_id: string | null;
  border_agent_id: string | null;
  active_operational_dataset: string | null;
}
```

- `mac_extended_address`: The MAC extended address of the device
- `extended_pan_id`: The extended PAN ID of the Thread network
- `border_agent_id`: The border agent ID

> - `active_operational_dataset`: The active operational dataset

Reviewer: ```suggestion
All payload properties describe the preferred Thread network which should be used when commissioning a Matter over Thread device.

- `mac_extended_address`: The MAC extended address 
- `extended_pan_id`: The extended PAN ID
- `border_agent_id`: The border agent ID
- `active_operational_dataset`: The active operational dataset (TLV)
```

##### `thread/import_credentials`

> Import Thread network credentials.

Reviewer: This one always makes me double check. It's importing from the server point of view. It's exporting from the mobile device point of view.

```suggestion
Import Thread network credentials from the device. Credentials should be sent using the [WebSocket API](/docs/api/websocket)
```

(unfortunately it looks like the command is undocumented so I cannot link to a specific heading at this point)

Payload: None

##### `improv/scan`

> Start scanning for Improv-enabled devices.

Reviewer: ```suggestion
Start scanning for Improv Wi-Fi devices.
```

Payload: None

##### `improv/configure_device`

> Configure a discovered Improv device.

Reviewer: ```suggestion
Configure a discovered Improv Wi-Fi device.
```

```ts
{
  url: string;
  muted: boolean;
}
```

- `url`: The HLS stream URL
- `muted`: Whether to start playback muted

##### `exoplayer/stop`

Stop HLS video playback.

Payload: None

##### `exoplayer/resize`

Resize the HLS video player.

Payload:

```ts
{
  left: number;
  top: number;
  right: number;
  bottom: number;
}
```

- `left`: Left coordinate of the player
- `top`: Top coordinate of the player
- `right`: Right coordinate of the player
- `bottom`: Bottom coordinate of the player

##### `entity/add_to`

Add an entity to a platform-specific location (e.g., homescreen widget).

Payload:

```ts
{
  entity_id: string;
  app_payload: string;
}
```

- `entity_id`: The entity to add
- `app_payload`: Opaque string received from `entity/add_to/get_actions`

##### `thread/store_in_platform_keychain`

Store Thread credentials in the platform keychain.

Payload:

```ts
{
  mac_extended_address: string | null;
  border_agent_id: string | null;
  active_operational_dataset: string;
  extended_pan_id: string;
}
```

- `mac_extended_address`: The MAC extended address of the device
- `border_agent_id`: The border agent ID
- `active_operational_dataset`: The active operational dataset

> - `extended_pan_id`: The extended PAN ID of the Thread network

Reviewer: Align with the previous description

```suggestion
- `mac_extended_address`: The MAC extended address
- `border_agent_id`: The border agent ID
- `active_operational_dataset`: The active operational dataset (TLV)
- `extended_pan_id`: The extended PAN ID
```

##### `focus_element`

Focus a specific element in the frontend.

Payload:

```ts
{
  element_id: string;
}
```

- `element_id`: The ID of the element to focus

> ### Messages from app to frontend

Reviewer: This list should be sorted a-z as well

> These messages are sent from the external app to the Home Assistant frontend. All messages expect a response.

Reviewer: What kind of response?

##### `restart`

Request the frontend to restart.

Payload: None

##### `navigate`

Navigate to a specific path in the frontend.

Payload:

```ts
{
  path: string;
  options?: {
    replace: boolean;
  };

> }

Reviewer: Can you describe the options?

 - is `path` relative or absolute
 - what does replace do

```

##### `notifications/show`

Show the notifications panel.

Payload: None

##### `sidebar/toggle`

Toggle the sidebar open/closed.

Payload: None

##### `sidebar/show`

Show the sidebar.

Payload: None

##### `automation/editor/show`

Open the automation editor.

Payload (optional):

```ts
{
  config?: Partial<AutomationConfig>;
}
```

##### `bar_code/scan_result`

> Send a barcode scan result to the frontend.

Reviewer: ```suggestion
Send a bar code scanner result to the frontend.
```

Payload:

```ts
{
  rawValue: string;
  format:
    | "aztec"
    | "code_128"
    | "code_39"
    | "code_93"
    | "codabar"
    | "data_matrix"
    | "ean_13"
    | "ean_8"
    | "itf"
    | "pdf417"
    | "qr_code"
    | "upc_a"
    | "upc_e"
    | "unknown";
}
```

- `rawValue`: A string decoded from the barcode data
- `format`: The barcode format as defined by the [Barcode Detection API](https://developer.mozilla.org/en-US/docs/Web/API/Barcode_Detection_API#supported_barcode_formats)

##### `bar_code/aborted`

Notify that barcode scanning was aborted.

Payload:

```ts
{
  reason: "canceled" | "alternative_options";
}
```

##### `improv/discovered_device`

> Notify that an Improv device was discovered.

Reviewer: ```suggestion
Notify that an Improv Wi-Fi device was discovered.
```

Payload:

```ts
{
  name: string;
}
```

##### `improv/device_setup_done`

> Notify that Improv device setup is complete.

Reviewer: ```suggestion
Notify that Improv Wi-Fi device setup is complete.
```

}
```

- `tag`: The tag ID to write
- `name`: The name of the tag as entered by the user, or `null` if no name has been set

##### `theme-update`

Notify the app that the theme has been updated. The app should refresh its status bar and navigation bar colors.

Payload: None

##### `thread/import_credentials`

> Import Thread network credentials from the device. Credentials should be sent using the [WebSocket API](/docs/api/websocket)

Reviewer: Forgot to include a `.` like all other descriptions in my previous suggestion, sorry.

```suggestion
Import Thread network credentials from the device. Credentials should be sent using the [WebSocket API](/docs/api/websocket).
```

#### Messages expecting a response

All commands in this section receive a result message back from the frontend:

```ts
{
  id: number;
  type: "result";
  success: boolean;
  result: null;
  error?: { code: string; message: string };
}
```

##### `automation/editor/show`

Open the automation editor to create a new automation. The config can be used to pre-fill the editor with specific triggers, conditions, and actions.

Payload (optional):

```ts
{
  config?: {
    alias?: string;
    description?: string;
    triggers?: Trigger | Trigger[];
    conditions?: Condition | Condition[];
    actions?: Action | Action[];
    mode?: "single" | "restart" | "queued" | "parallel";
    max?: number;
  };
}
```

- `config.alias`: Pre-filled name for the automation
- `config.description`: Pre-filled description
- `config.triggers`: One or more triggers to pre-populate
- `config.conditions`: One or more conditions to pre-populate
- `config.actions`: One or more actions to pre-populate
- `config.mode`: Execution mode for the automation
- `config.max`: Maximum number of concurrent runs (only applies to `queued` and `parallel` modes)

##### `bar_code/aborted`

> Notify that barcode scanning was aborted.

Reviewer: ```suggestion
Notify that bar code scanning was aborted.
```
